source-container-build/app/test_source_build.py

```diff
@@ -0,0 +1,1154 @@
+        it should be included as an extra source.
+        """
+        sib_dirs = SourceImageBuildDirectories()
+        self._mark_cachi2_has_run()
+        with open(os.path.join(self.cachi2_dir, "cachi2.env"), "w") as f:
+            f.write("no matter what the content is")
+
+        result = source_build.gather_prefetched_sources(self.work_dir, self.cachi2_dir, sib_dirs)
+
+        self.assertFalse(result)
+        self.assertListEqual([os.path.join(self.work_dir, "cachi2_env")], sib_dirs.extra_src_dirs)
+
+    def _test_gather_deps_by_package_manager(self, fetched_deps: list[str], dep_ext: str):
+        self._mark_cachi2_has_run()
+        create_fake_dep_packages(self.cachi2_output_dir, fetched_deps)
+
+        sib_dirs = SourceImageBuildDirectories()
+        result = source_build.gather_prefetched_sources(self.work_dir, self.cachi2_dir, sib_dirs)
+        self.assertTrue(result)
+
+        def _has_known_file_ext(filename: str) -> bool:
+            return (
+                filename.endswith(".tar.gz")
+                or filename.endswith(".tgz")
+                or filename.endswith(".zip")
+            )
+
+        # Remove the noise introduced by test on gomod package manager
+        deps_with_known_file_ext = [item for item in fetched_deps if _has_known_file_ext(item)]
+
+        prefetched_sources_dir = os.path.join(self.work_dir, "prefetched_sources")
+        # Check constructed directories: work_dir/prefetched_sources/src-N
+        expected_src_dirs = [
+            os.path.join(prefetched_sources_dir, f"src-{count}")
+            for count in range(len(deps_with_known_file_ext))
+        ]
+        self.assertListEqual(sorted(expected_src_dirs), sorted(sib_dirs.extra_src_dirs))
+
+        gathered_deps = []  # collect the dep packages gathered by the method
+        for dir_path, subdir_names, file_names in os.walk(prefetched_sources_dir):
+            for name in file_names:
+                if name.endswith(dep_ext):
+                    gathered_deps.append(name)
+
+        self.assertListEqual(
+            sorted([os.path.basename(dep) for dep in deps_with_known_file_ext]),
+            sorted(gathered_deps),
+        )
+
+    def test_gather_pip_deps(self):
+        pip_deps = ["pip/requests-1.0.0.tar.gz", "pip/Flask-1.2.3.tar.gz"]
+        self._test_gather_deps_by_package_manager(pip_deps, ".tar.gz")
+
+    def test_gather_npm_deps(self):
+        npm_deps = ["npm/has-1.0.3.tgz", "npm/express-4.18.1.tgz", "npm/bytes-3.1.2.tgz"]
+        self._test_gather_deps_by_package_manager(npm_deps, ".tgz")
+
+    def test_gather_go_deps(self):
+        gomod_deps = [
+            "gomod/pkg/mod/cache/download/gopkg.in/yaml.v2/@v/v2.4.0.zip",
+            "gomod/pkg/mod/cache/download/gopkg.in/yaml.v2/@v/v2.4.0.info",
+            "gomod/pkg/mod/cache/download/github.com/go-logr/logr/@v/v1.2.3.lock",
+        ]
+        self._test_gather_deps_by_package_manager(gomod_deps, ".zip")
```

Comment:
Maybe these 3 should just be one test? I think it would simplify the test

And we want source-build to work when all the package managers are present at the same time, so IMO it makes sense to test it like that

Reply:
I'd like to keep these 3 separate cases so that the test shows the specific cases explicitly that the implementation should work.

A new test is added for the the case of "all the package managers are present at the same time".